src/System.Private.CoreLib/shared/System/Resources/ResourceReader.cs

```diff
@@ -743,7 +758,81 @@ private object _LoadObjectV2(int pos, out ResourceTypeCode typeCode)
+            return DeserializeObject(typeIndex);
+        }
+
+        // Helper method to deserialize a type
+        private Object DeserializeObject(int typeIndex)
+        {
+            if (!_permitDeserialization)
+            {
+                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
+            }
+
+            if (_binaryFormatter == null)
+            {
+                InitializeBinaryFormatter();
+            }
+
+            Type type = FindType(typeIndex);
+ 
+            // Ensure that the object we deserialized is exactly the same
+            // type of object we thought we should be deserializing. 
+            Object graph;
+            graph = s_deserializeMethod(_binaryFormatter, _store.BaseStream);
+            
+            // This check is about correctness, not security at this point.
+            if (graph.GetType() != type)
+                throw new BadImageFormatException(SR.Format(SR.BadImageFormat_ResType_SerBlobMismatch, type.FullName, graph.GetType().FullName));
+ 
+            return graph;
+        }
+
+        private void InitializeBinaryFormatter()
+        {
+            if (s_binaryFormatterType == null)
+            {
+                Type binaryFormatterType = Type.GetType(
+                    "System.Runtime.Serialization.Formatters.Binary.BinaryFormatter, System.Runtime.Serialization.Formatters, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
+                    throwOnError: false);
+                if (binaryFormatterType != null)
+                {
+                    ConstructorInfo binaryFormatterCtor = binaryFormatterType.GetConstructor(new Type[0]);
+                    MethodInfo binaryFormatterDeserialize = binaryFormatterType.GetMethod("Deserialize", new Type[] { typeof(Stream) });
+
+                    s_binaryFormatterType = binaryFormatterType;
+                    // create an unbound delegate that can accept a BinaryFormatter instance as object
+                    s_deserializeMethod = (Func<object, Stream, object>)typeof(ResourceReader)
+                                           .GetMethod(nameof(CreateUntypedDelegate), BindingFlags.NonPublic | BindingFlags.Static)
+                                           .MakeGenericMethod(binaryFormatterType)
+                                           .Invoke(null, new object[] { binaryFormatterDeserialize });
+                }
+                else
+                {
+                    // set a sentinel to indicate that BinaryFormatter wasn't found
+                    s_binaryFormatterType = typeof(object);
+                    s_deserializeMethod = null;
+                }
+            }
+
+            if (s_binaryFormatterType != null && s_deserializeMethod == null)
+            {
+                // we couldn't find the type
+                throw new NotSupportedException(SR.NotSupported_ResourceObjectSerialization);
```

Comment:
If we can't find the type why are we throwing a NotSupportedException? I assume the type will always be found unless System.Runtime.Serialization.Formatters.dll is missing in the shared folder which indicates a bad environment. Shouldn't we throw a FileNotFoundException instead?

Reply:
> Shouldn't we throw a FileNotFoundException instead?

It is what we are doing in other similar cases. We just let the `Type.GetType` throw the exception. `ILGenerator.GetConsoleType` is one of many examples.

Reply:
I'm trying to preserve the behavior that we had in nca1.x-2.x rather than add a new exception.  Its a valid scenario that BinaryFormatter is missing: a self-contained application that has been trimmed.

Reply:
I do not think we need to maintain compatibility for the trimmed applications here.

It is better to throw FileNotFoundException when the application is trimmed too much. Converting the 
 FileNotFoundException into NotSupportedException just makes it harder to diagnose bugs caused by too aggressive trimming.